jans-cedarling/cedarling/src/common/cedar_schema/cedar_json/action.rs

```diff
@@ -5,27 +5,216 @@
  * Copyright (c) 2024, Gluu, Inc.
  */
 
+use crate::{
+    authz::entities::CEDAR_POLICY_SEPARATOR, common::cedar_schema::cedar_json::SchemaDefinedType,
+};
+
 use super::{
     entity_types::{
         CedarSchemaEntityAttribute, CedarSchemaEntityType, PrimitiveType, PrimitiveTypeKind,
     },
-    CedarSchemaRecord,
+    CedarSchemaEntities, CedarSchemaJson, CedarSchemaRecord, CedarType, GetCedarTypeError,
 };
 use serde::{de, ser::SerializeMap, Deserialize, Serialize};
 use serde_json::{json, Value};
 use std::collections::{HashMap, HashSet};
 
-pub type EntityRef = String;
+type AttrName = String;
+
+#[derive(Debug, Eq, Hash, PartialEq)]
+pub struct CtxAttribute {
+    pub namespace: String,
+    pub key: String,
+    pub kind: CedarType,
+}
+
+pub struct Action<'a> {
+    pub principal_entities: HashSet<String>,
+    pub resource_entities: HashSet<String>,
+    pub context_entities: Option<HashSet<CtxAttribute>>,
+    pub schema_entities: &'a CedarSchemaEntities,
+    pub schema: &'a ActionSchema,
+}
+
+impl Action<'_> {
+    /// Builds the JSON representation of context entities for a given action.
+    ///
+    /// This method processes the context attributes of the action and generates a
+    /// corresponding JSON value. The context may include entity references (with
+    /// `type` and `id`) and other values, which can be mapped through the provided
+    /// `id_mapping` and `value_mapping`.
+    ///
+    /// The `id_mapping` param is a A `HashMap` that maps context attribute keys
+    /// (like `"access_token"`) to their corresponding `id`s (like `"acs-tkn-1"`).
+    ///
+    /// # Usage Example
+    ///
+    /// ```rs
+    /// let id_mapping = HashMap::from([("access_token".to_string(), "acs-tkn-1".to_string())]);
+    /// let json = action.build_ctx_entities_json(id_mapping, value_mapping);
+    /// ```
+    pub fn build_ctx_entity_refs_json(
+        &self,
+        id_mapping: HashMap<String, String>,
+    ) -> Result<Value, BuildJsonCtxError> {
+        let mut json = json!({});
+
+        if let Some(ctx_entities) = &self.context_entities {
+            for attr in ctx_entities.iter() {
+                if let CedarType::TypeName(type_name) = &attr.kind {
+                    let id = match id_mapping.get(&attr.key) {
+                        Some(val) => val,
+                        None => Err(BuildJsonCtxError::MissingIdMapping(attr.key.clone()))?,
+                    };
+                    let type_name = [attr.namespace.clone(), type_name.to_string()]
+                        .join(CEDAR_POLICY_SEPARATOR);
+                    json[attr.key.clone()] = json!({"type": type_name, "id": id});
```

Comment:
we can avoid clonning

```suggestion
                    let type_name =
                        [attr.namespace.as_str(), type_name.as_str()].join(CEDAR_POLICY_SEPARATOR);
```

Reply:
fixed here: [7133729](https://github.com/JanssenProject/jans/pull/10387/commits/71337294dd25ad42d58efee148f87d63b6010d1f)